src/Driver/PDO/Connection.php

```diff
@@ -87,17 +91,80 @@ public function quote(string $value): string
         return $this->connection->quote($value);
     }
 
-    public function lastInsertId(?string $name = null): string
+    /**
+     * {@inheritDoc}
+     */
+    public function lastInsertId()
     {
+        $driverName = $this->connection->getAttribute(PDO::ATTR_DRIVER_NAME);
+
+        if ($driverName === 'oci') {
+            throw IdentityColumnsNotSupported::new();
+        }
+
         try {
-            if ($name === null) {
-                return $this->connection->lastInsertId();
+            $lastInsertId = $this->connection->lastInsertId();
+        } catch (PDOException $exception) {
+            [$sqlState] = $exception->errorInfo;
+
+            // PDO PGSQL throws a 'lastval is not yet defined in this session' error when no identity value is
+            // available, with SQLSTATE 55000 'Object Not In Prerequisite State'
+            if ($sqlState === '55000' && $this->connection->getAttribute(PDO::ATTR_DRIVER_NAME) === 'pgsql') {
+                throw NoIdentityValue::new($exception);
             }
 
-            return $this->connection->lastInsertId($name);
+            throw Exception::new($exception);
+        }
+
+        if (! $this->isValidLastInsertIdOrSequenceNumber($lastInsertId)) {
+            throw NoIdentityValue::new();
+        }
+
+        return $lastInsertId;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getSequenceNumber(string $name)
+    {
+        $driverName = $this->connection->getAttribute(PDO::ATTR_DRIVER_NAME);
+
+        if ($driverName === 'mysql') {
+            throw SequencesNotSupported::new();
+        }
+
+        if ($driverName === 'oci') {
+            throw new SequencesNotSupported('Oracle supports sequences, but the PDO OCI driver does not support them.');
+        }
+
+        try {
+            $sequenceNumber = $this->connection->lastInsertId($name);
         } catch (PDOException $exception) {
+            [$sqlState] = $exception->errorInfo;
+
+            // PDO PGSQL throws an 'Undefined table' when the sequence does not exist
+            if ($sqlState === '42P01' && $this->connection->getAttribute(PDO::ATTR_DRIVER_NAME) === 'pgsql') {
+                throw SequenceDoesNotExist::new($name, $exception);
+            }
```

Comment:
I'm not comfortable with having these checks in the driver Connection rather than in the `ExceptionConverter`, but we need some context when converting to an exception, as `42P01` can either mean a `TableNotFoundException` or a `SequenceDoesNotExist`.

Reply:
So are you suggesting to move this error handling to the exception converter or to leave it here?

Reply:
If we want to move it to the exception converter, we have to add a bit of context to the API, but I am not sure in which form: name of the calling method? An enum value like `QUERY` / `LAST_INSERT_ID` / `GET_SEQUENCE_NUMBER`? Or just leave this code here, but it's not pretty. TBH I don't know what's the best approach, or even if there is one.

Reply:
I think it's good enough to leave it there for now; adding complexity / extra parameters to the ExceptionConverter is a bit premature IMO, and may do more harm than good.